src/MarginTrading.MarketMaker/Services/ExtPrices/Implementation/GenerateOrderbookService.cs

```diff
@@ -87,73 +86,81 @@ public Orderbook OnNewOrderbook(ExternalOrderbook orderbook)
             {
-                Trace.Write(assetPairId + " warn trace", $"Skipping not configured exchange {orderbook.ExchangeName} for {assetPairId}");
-                return (null, null);
+                Trace.Write(TraceLevelGroupEnum.WarnTrace, assetPairId,
+                    $"Skipping not configured exchange {orderbook.ExchangeName}",
+                    new {Event = "NotConfiguredExchangeSkipped", orderbook.ExchangeName});
+                return (null, null, "Skipping not configured exchange");
             }
 
             var allOrderbooks = _orderbooksService.AddAndGetByAssetPair(orderbook);
             var now = orderbook.LastUpdatedTime;
             var (exchangesErrors, validOrderbooks) = MarkExchangesErrors(assetPairId, allOrderbooks, now);
-            var primaryExchange = _primaryExchangeService.GetPrimaryExchange(assetPairId, exchangesErrors, now, orderbook.ExchangeName);
+            var primaryExchange =
+                _primaryExchangeService.GetPrimaryExchange(assetPairId, exchangesErrors, now, orderbook.ExchangeName);
             if (primaryExchange == null)
             {
-                return (null, null);
+                return (null, null, "No primary exchange");
             }
 
             if (primaryExchange != orderbook.ExchangeName)
             {
-                return (null, primaryExchange);
+                return (null, primaryExchange, "Orderbook not from primary exchange");
             }
 
             if (!allOrderbooks.TryGetValue(primaryExchange, out var primaryOrderbook))
             {
-                _log.WriteWarningAsync(nameof(GenerateOrderbookService), null,
-                    $"{primaryExchange} not found in allOrderbooks ({allOrderbooks.Keys.ToJson()})");
-                return (null, primaryExchange);
+                _log.WriteErrorAsync(nameof(GenerateOrderbookService), null,
+                    new Exception($"{primaryExchange} not found in allOrderbooks ({allOrderbooks.Keys.ToJson()})")
+                    {
+                        Data = {{"AssetPairId", assetPairId}}
+                    });
+                return (null, primaryExchange, "Primary exchange orderbook not found");
             }
 
             _stopTradesService.FinishCycle(primaryOrderbook, now);
             var resultingOrderbook = Transform(primaryOrderbook, validOrderbooks);
-            if (!CanSendOrderbook(resultingOrderbook))
+            if (TryFindSkipOrderbookReason(resultingOrderbook) is string reason)
             {
-                return (null, primaryExchange);
+                return (null, primaryExchange, reason);
             }
 
-            return (resultingOrderbook, primaryExchange);
+            return (resultingOrderbook, primaryExchange, null);
         }
 
-        private bool CanSendOrderbook(Orderbook orderbook)
+        [CanBeNull]
+        private string TryFindSkipOrderbookReason(Orderbook orderbook)
```

Comment:
This method looks very strange...

Reply:
Please tell me in detail why you think it's strange?